tests/csapi/apidoc_room_event_test.go

```diff
@@ -0,0 +1,206 @@
+package csapi_tests
+
+import (
+	"fmt"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/match"
+)
+
+func stringSliceContains(slice []string, s string) bool {
+	for _, value := range slice {
+		if value == s {
+			return true
+		}
+	}
+	return false
+}
+
+// sytest: Events come down the correct room
+func TestEventsInCorrectRoom(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintAlice)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+
+	const (
+		roomAmount     = 10
+		goConcurrency  = 4
+		channelTimeout = 10 * time.Second
+	)
+
+	// Create all rooms
+
+	var doneRooms = make(chan string, roomAmount)
+	var todoRooms = make(chan bool, goConcurrency)
+
+	for i := 0; i < goConcurrency; i++ {
+		go func() {
+			for {
+				_, ok := <-todoRooms
+
+				if !ok {
+					return
+				}
+
+				roomId := alice.CreateRoom(t, map[string]interface{}{
+					"preset": "public_chat",
+				})
+
+				doneRooms <- roomId
+			}
+		}()
+	}
+
+	for i := 0; i < roomAmount; i++ {
+		todoRooms <- true
+	}
+
+	close(todoRooms)
+
+	var rooms []string
+
+	for i := 0; i < roomAmount; i++ {
+		select {
+		case newRoom := <-doneRooms:
+			rooms = append(rooms, newRoom)
+		case <-time.After(channelTimeout):
+			t.Fatalf("Room creation timed out, got to %d rooms (from %d)", i, roomAmount)
+		}
+	}
+
+	// Send events to all rooms with corresponding roomID
+
+	// get current next_batch
+	_, since := alice.MustSync(t, client.SyncReq{TimeoutMillis: "0"})
+
+	type txnAndId struct {
+		txn    int
+		roomId string
+	}
+
+	var doneEvents = make(chan string, roomAmount)
+	var todoEvents = make(chan txnAndId, goConcurrency)
+
+	txnCount := 0
+
+	for i := 0; i < goConcurrency; i++ {
+		go func() {
+			for {
+				tid, ok := <-todoEvents
+
+				if !ok {
+					return
+				}
+
+				paths := []string{"_matrix", "client", "r0", "rooms", tid.roomId, "send", "m.room.message", strconv.Itoa(tid.txn)}
+				res := alice.MustDoFunc(t, "PUT", paths, client.WithJSONBody(t, map[string]interface{}{
+					"msgtype": "m.text",
+					"body":    tid.roomId,
+				}))
+				body := client.ParseJSON(t, res)
+				eventID := client.GetJSONFieldStr(t, body, "event_id")
+
+				doneEvents <- eventID
+			}
+		}()
+	}
+
+	for _, id := range rooms {
+		todoEvents <- txnAndId{
+			roomId: id,
+			txn:    txnCount,
+		}
+
+		txnCount++
```

Comment:
You might as well just have `for i, id := range rooms` and then inject `i` rather than using a new variable `txnCount` and have to increment it.